File: AoC/aoc-2024/day1.js
import { readFile } from 'fs/promises';

const parse = (input) => {
    let trimmedInput = input.trimEnd();
    let lines = trimmedInput.split("\n");
    let nums = [];
    for (let i = 0; i < lines.length; i++) {
        let splitline = lines[i].split("   ");
        nums.push(splitline[0]);
        nums.push(splitline[1]);
    }

    let leftList = [];
    let rightList = [];

    for (let i = 0; i < nums.length; i++) {
        if (i % 2 === 0) {
            leftList.push(nums[i])
        } else {
            rightList.push(nums[i])
        }
    }

    return [leftList, rightList]
}

const getOccurrences = (arr, item) => {
    var n = 0;
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] === item) { n++ }
    }
    return n;
}

// Read and parse input
let input = await readFile("/Users/littletheo/non-git/aoc-2024/day1input.txt", "utf8");
let [leftList, rightList] = parse(input);

// Sort and count dist btwn leftList and rightList
leftList.sort((a, b) => a - b);
rightList.sort((a, b) => a - b);

let total = 0;

for (let i = 0; i < leftList.length; i++) {
    let dist = Math.abs(leftList[i] - rightList[i]);
    total += dist;
}

// Count similarity btwn leftList and rightList
let simScore = 0;
for (let i = 0; i < leftList.length; i++) {
    let numOccurrences = getOccurrences(rightList, leftList[i]);
    simScore += leftList[i] * numOccurrences;
}

console.log(simScore);

File: AoC/aoc-2024/day1.py
def parse_input(input):
    left_list = []
    right_list = []
    i = 0
    for id in input.split():
        if i % 2 == 0:
            left_list.append(int(id))
        else:
            right_list.append(int(id))

        i += 1

    return left_list, right_list

def main():
    with open('day1input.txt', 'r') as input:
        left_list, right_list = parse_input(input.read())

    sorted_left = sorted(left_list)
    sorted_right = sorted(right_list)
    total = 0

    for i in range(0, len(sorted_left)):
        total += abs(sorted_left[i] - sorted_right[i])

    return total

if __name__ == '__main__':
    response = main()
    print(response)

